primap2/pm2io/_data_reading.py

```diff
@@ -41,6 +41,178 @@
+    time_format: str = "%Y-%m-%d",
+) -> pd.DataFrame:
+    """convert a DataFrame in long (tidy) format into the PRIMAP2 interchange format.
+
+    Columns can be renamed or filled with default values to match the PRIMAP2 structure.
+    Where we refer to "dimensions" in the parameter description below we mean the basic
+    dimension names without the added terminology (e.g. "area" not "area (ISO3)"). The
+    terminology information will be added by this function. You can not use the short
+    dimension names in the attributes (e.g. "cat" instead of "category").
+
+    Parameters
+    ----------
+    data_long: str, pd.DataFrame
+        Long format DataFrame file which will be converted.
+
+    coords_cols : dict
+        Dict where the keys are column names in the files to be read and the value is
+        the dimension in PRIMAP2. To specify the data column containing the observable,
+        use the "data" key. For secondary categories use a ``sec_cats__`` prefix.
+
+    coords_defaults : dict, optional
+        Dict for default values of coordinates / dimensions not given in the csv files.
+        The keys are the dimension names and the values are the values for
+        the dimensions. For secondary categories use a ``sec_cats__`` prefix.
+
+    coords_terminologies : dict
+        Dict defining the terminologies used for the different coordinates (e.g. ISO3
+        for area). Only possible coordinates here are: area, category, scenario,
+        entity, and secondary categories. For secondary categories use a ``sec_cats__``
+        prefix. All entries different from "area", "category", "scenario", "entity", and
+        ``sec_cats__<name>`` will raise a ValueError.
+
+    coords_value_mapping : dict, optional
+        A dict with primap2 dimension names as keys. Values are dicts with input values
+        as keys and output values as values. A standard use case is to map gas names
+        from input data to the standardized names used in primap2.
+        Alternatively a value can also be a function which transforms one CSV metadata
+        value into the new metadata value.
+        A third possibility is to give a string as a value, which defines a rule for
+        translating metadata values. For the "category", "entity", and "unit" columns,
+        the rule "PRIMAP1" is available, which translates from PRIMAP1 metadata to
+        PRIMAP2 metadata.
+
+    coords_value_filling : dict, optional
+        A dict with primap2 dimension names as keys. These are the target columns where
+        values will be filled (or replaced). Vales are dicts with primap2 dimension names
+        as keys. These are the source columns. The values are dicts with source value -
+        target value mappings.
+        This can be used to e.g. fill missing category codes based on category names or
+        to replace category codes which do not meet the terminology using the category
+        names.
```

Comment:
I think I don't understand what `coords_value_filling` can do what `coords_value_mapping` can't do. I think that should be spelled out in the documentation

Reply:
OK, will clarify that. 
The idea is that sometimes you have non-unique category codes but unique descriptions. This can only be fixed by using information from one column on another column. But coords_value_filling can actually replace coords_value_mapping. Just that the format is a bit more complicated.

Reply:
It's a functionality that was missing in PRIMAP1 and complicated data reading in some cases

Reply:
added a bit of documentation where the function is used. We could integrate it also into other reading functions. And we have to decide if we replace `coords_value_mapping`.

Reply:
Cool, now I understand the functionality. I like the possibility, because this is already what is needed for units sometimes (because you also need the entity) and so far there is a hack which makes it possible for built-in coord mappings like `PRIMAP1`, but not flexibly for coord mappings supplied by the user. Would have to think how we can preserve the functionality of `coords_value_mapping` without making everything too complicated.

Reply:
Currently information from different columns can not be combined, which is necessary for the PRIMAP1 to PRIMAP2 unit conversion.

Reply:
Hmm... I'm thinking about a somewhat nice API which is not completely confusing...

Using introspection, function pointers could be used for a quite nice API, like this:

```python
def calculate_unit(entity, unit, category):
    """just a silly example for a complex multi-category mapping"""
    if entity == "KYOTOGHG":
        return "CO2"
    elif unit == "kt" and category == "average speed of international marine trade":
        return "knots"
    else:
        return unit

coords_value_mapping={"unit": calculate_unit})
```
Where the function can then extract the needed inputs from the parameter names (in the example, `unit`, `entity`, and `category`). Would be kind of straight forward and easy to understand without deeply nested dictionaries. Does that make sense? Or is it too verbose for the "I just want to fill one column from another column" use case?

Reply:
the problem I see here is that you need harmonized column names to re-use the function. But if it's called in the right place (directly before conversion to IF) and after renaming this should be possible (when using the column name aliases as in the xarray format). I don't think it's too confusing and one could still use the old dict input, or provide a template for a function that does the simple replacing